extensions/vault/model/src/main/java/io/quarkus/vault/runtime/client/dto/pki/VaultPKIConstants.java

```diff
@@ -0,0 +1,8 @@
+package io.quarkus.vault.runtime.client.dto.pki;
+
+public class VaultPKIConstants {
+
+    static final String CERTIFICATE_FORMAT = "pem";
```

Comment:
it was something I missed.
so what you are doing here is defining defaults for the certificate format and key format so you get consistency across all the operations. 
I would do 2 things:
- rename variables using `DEFAULT_CERTIFICATE_FORMAT` and `DEFAULT_KEY_FORMAT`
- I would add a sentence in your (very good) documentation to expose those choices (i.e.: _the default certificate format is pem, and the default private key format is pkcs8_)

optionally, I would create 2 enums:
- certificate format: pem, der, or pem_bundle
- private key format: der, pkcs8

I am not seeing a way to change the value on `VaultPKIGenerateCertificateBody.format` and `privateKeyFormat`. now I understand better https://github.com/quarkusio/quarkus/pull/18636#discussion_r672813701

there are not just defaults. they are the only supported options. 

in terms of api, there were different options: creating an interface or abstract class, and concrete subclasses for the representations in each format, or a single class (e.g. `CertificateData` with a `format` field, and some specific fields `String pemData; byte[] derData`, with appropriate getters to retrieve data, or test for type).

if you look at the spring [api](https://docs.spring.io/spring-vault/docs/current/api/org/springframework/vault/core/VaultPkiOperations.html) you can see they return a [CertificateBundle](https://docs.spring.io/spring-vault/docs/current/api/org/springframework/vault/support/CertificateBundle.html), which extends a `Certificate` class providing `getCertificate()` that returns a `String` (PEM or DER - I guess that is your option 2. I am surprised they do not have a `formatType` field), and a convenient `getX509Certificate()` if the format is DER, which will return the `java.security` object.
They also have a `getPrivateKeySpec()` on the bundle that returns the `java.security.spec.KeySpec`, which is an interface representing the different type and encoding flavors. e.g. [PKCS8EncodedKeySpec](https://docs.oracle.com/javase/8/docs/api/java/security/spec/PKCS8EncodedKeySpec.html)

don't you think that some people will want to use der or pem_bundle for the certificate format, or der for the private key format?

so basically I see 3 options:
- do nothing: opiniated choice. we believe there is no case for `der`.
- in `getCertificate` (and the same logic would apply in other places) create a class to represent a `Certificate` with a `format` field, and leave the `String` type for the `certificate`, `issuingCA`, ... if using DER, these will be the bytes encoded in base64 (exactly as vault does).
- same as above option, but introducing a class hierarchy to model the different formats.

option 2 might be a good compromise. what do you think?

Reply:
Implemented via `CertificateData`, `CSRData` & `PrivateKeyData` interfaces. Each has a `DER` and `PEM` implementation for respective formats.

For certificates the format is selectable by setting the `format` option on the associated options object. The `CertificateData` interface has a `getCertificate()` that will parse its data into an `X509Certificate`; this is supported for both PEM and DER formats.

Due to the vagaries of the Vault api, I didn't expose a format for private key, because Vault determines the format of the private key by combining the `format` and `private_key_format` fields. This means that for  `PrivateKeyData` the format is selectable by the same `format` as for certificates but when necessary there's an added `privateKeyEncoding` field on the associated options objects. The field is an enum with `RAW` and `PKCS8` and it only determines if the field is wrapped in PKCS8 packaging. The `PrivateKeyData` interface has a `getKeySpec()` that returns a `KeySpec` for the key. Due to the limited support in Java's security API for EC keys this only works on PKCS8 encoded keys and throws and exception if called on `RAW` encoded keys.

Reply:
Note that some of the API's cannot actually return DER encodings (e.g. `getCertificate()`) and others (e.g. `getCertificateAuthority()`) would require adding a bunch of machinery to return and parse binary responses (that are not in a normal Vault response format).